### PR TITLE
Takeaway challenge - Conor Butler

### PR DESCRIPTION
**Pull request by ConorButler**
Would love to have tried the bonus step of ordering with a text message but can't get sinatra to setup on my environment without some problems.

### CODE REVIEW

File: lib/order.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+class Order
```
Comment (t-kellett) [inline lib/order.rb:1]: There are some (useful) methods in here that I don't think are specified in the spec (i.e. being able to remove and clear the basket). I understand why you would implement them but in real world projects I would think you possibly miss delivering against the real spec by going beyond the scope to deliver things you think should be delivered.

File: lib/takeaway.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  attr_reader :current_order
+  @@sms_client = SMS.new
+
+  def initialize(dishes, order_class: Order)
+    dishes = [dishes] unless dishes.is_a? Array # can pass in 1 or more dishes
+    @dishes = dishes
+    @order_class = order_class
+  end
+
+  def dishes
```
Comment (t-kellett) [inline lib/takeaway.rb:14]: Unclear what this is used for

File: lib/takeaway.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    @dishes.dup
+  end
+
+  def menu
+    puts dishes.map { |dish| format(dish) }.join("\n\n")
+  end
+
+  def select_dish(dish, quantity = 1)
```
Comment (t-kellett) [inline lib/takeaway.rb:22]: very nice, concise code

File: spec/takeaway_spec.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
+require 'takeaway'
```
Comment (t-kellett) [inline spec/takeaway_spec.rb:1]: Clearly defined tests, nice

**Review by t-kellett — COMMENTED** [review]
Generally really nice implementation. Tests are very clear and the context blocks make it easy to understand what you're testing.

To improve I think there are 2 things that could be changed:
- A customer interface to interact with the takeaway - currently to place an order you need to know the code intimately and a real-world customer is unlikely to know this
- There are some methods here not described in the spec, which while provide useful functionality, may be rejected by a customer as something they have not asked for.

File: README.md [hunk 5]
```diff
@@ -47,37 +16,48 @@ So that I am reassured that my order will be delivered on time
...
+After struggling to get both to work, I tried to mock it as best as possible with just the existing methods. It currently tests that the string passed in with the **.text** method in Takeaway is successfully sent as the **:body** and received by a mock client. I'm not sure how effective this test is.
 
-Notes on Test Coverage
-------------------
+The other tests account for some edge cases but not all, as there were some I was not sure how to handle. The program supports functionality with multiple **Takeaway** objects, each with their own selection of **dishes**, however in theory you could pass the same **Dish** object into multiple takeaways (let's say takeaway1 and takeaway2), and then order this dish in takeaway1's instance from takeaway2.
 
-You can see your [test coverage](https://github.com/makersacademy/course/blob/main/pills/test_coverage.md) when you run your tests.
+I couldn't make an efficient solution in the time allocated. I think it would probably need a new Class that stores all known takeaways to check if each set of Dishes is unique, perhaps using some sort of id number.
```
Comment (JohnForster) [inline README.md:63]: Excellent Readme, well done!

File: lib/order.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  def clear_basket
+    @basket = []
+  end
+
+  def delivery_time
+    (Time.now + 3600).strftime("%H:%M")
+  end
+end
```
Comment (JohnForster) [inline lib/order.rb:31]: Good use of SRP, these methods are very clean.

File: lib/takeaway.rb [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+class Takeaway
+  attr_reader :current_order
+  include Menu
+  
+  @@sms_client = SMS.new
+
+  def initialize(dishes, order_class: Order)
+    dishes = [dishes] unless dishes.is_a? Array # can pass in 1 or more dishes
+    @dishes = dishes
```
Comment (JohnForster) [inline lib/takeaway.rb:19]: Could this be more descriptively named? Are the dishes chosen by the user or are they the dishes available?

File: lib/takeaway.rb [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  end
+
+  def dish_check(dish)
+    raise "Please select a dish from this takeaway." unless @dishes.include?(dish)
+  end
+
+  def order_check
+    raise "Please make an order first" if @current_order.nil?
+  end
+end
```
Comment (JohnForster) [inline lib/takeaway.rb:73]: There seems to be some overlap between this `Takeaway` class and your `Order` class. Can you summarise what each of them is for? Does every method fit that summary?

File: spec/takeaway_spec.rb [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+      other_dish = double("Other Dish")
+      expect { takeaway.select_dish(other_dish) }.to raise_error("Please select a dish from this takeaway.")
+    end
+  end
+end
+
+def add_1_order(dish)
+  expect(order).to receive(:add).with(dish) # Dish class handles the actual adding and removing
+  takeaway.select_dish(dish)
+end
```
Comment (JohnForster) [inline spec/takeaway_spec.rb:87]: Good mocking, well done.

File: spec/takeaway_spec.rb [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    end
+
+    it "confirms order with a text" do
+      takeaway.confirm_order
+    end
+
+    it "resets the current order after confirming" do
+      takeaway.confirm_order
+      expect(takeaway.current_order).to eq(nil)
+    end
```
Comment (JohnForster) [inline spec/takeaway_spec.rb:69]: You're testing the internal state of the class here, rather than the behaviour. Why is current_order exposed publicly?
